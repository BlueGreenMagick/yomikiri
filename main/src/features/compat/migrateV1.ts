import { VERSION } from "@/consts";
import type {
  AnkiNoteV1,
  AnkiTemplateFieldContentV1,
  AnkiTemplateFieldSentenceOptionsV1,
  AnkiTemplateFieldTypesV1,
  AnkiTemplateFieldV1,
  AnkiTemplateFieldWordOptionsV1,
  AnkiTemplateV1,
  FieldV1,
  StoredConfiguration1V1,
  StoredConfiguration2V1,
  StoredConfiguration3V1,
  StoredConfigurationV1,
} from "./types/typesV1";
import type { StoredConfigurationV2 } from "./types/typesV2";

export interface MigrateV1Props {
  config: StoredConfigurationV1;
}

export interface MigrateV1Result {
  config: StoredConfigurationV2;
}

export function migrateV1({ config }: MigrateV1Props): MigrateV1Result {
  if (
    config.config_version !== undefined &&
    config.config_version > 3
  ) {
    throw new Error(
      `Expected config_version to be less than 3, but received: ${config.config_version}`,
    );
  } else if (config.config_version === 3) {
    return { config };
  } else if (
    config.config_version === undefined &&
    config.version === undefined
  ) {
    const config = {
      config_version: 3,
      version: VERSION,
    } as const;
    return { config };
  }

  if (config.config_version === undefined) {
    console.debug(`Migrating config object from v1 to v2`);
    config = migrateConfig1(config);
  }
  if (config.config_version === 2) {
    console.debug(`Migrating config object from v2 to v3`);
    config = migrateConfig2(config);
  }

  return { config };
}

function migrateConfig1(config: StoredConfiguration1V1): StoredConfiguration2V1 {
  return {
    ...config,
    config_version: 2,
  };
}

function migrateConfig2(
  config: StoredConfiguration2V1,
): StoredConfiguration3V1 {
  const newConfig = {
    ...config,
    config_version: 3,
  } as StoredConfiguration3V1;

  const ankiTemplate = config["anki.template"];
  if (ankiTemplate !== undefined && ankiTemplate !== null) {
    newConfig["anki.anki_template"] = migrateAnkiTemplate_2(ankiTemplate);
  }

  return newConfig;
}

function migrateAnkiTemplate_2(template: AnkiNoteV1): AnkiTemplateV1 {
  return {
    ...template,
    fields: template.fields.map(fieldTemplateToAnyFieldTemplate),
  };
}

export function fieldTemplateToAnyFieldTemplate(fld: FieldV1): AnkiTemplateFieldV1 {
  const type = fld.value;
  const name = fld.name;
  if (type === "") {
    return {
      name,
      content: type,
    };
  } else if (["word", "word-furigana", "word-kana"].includes(type)) {
    const options: AnkiTemplateFieldWordOptionsV1 = {
      form: "as-is",
      style: type === "word-furigana" ?
        "furigana-anki" :
        type === "word-kana" ?
        "kana-only" :
        "basic",
    };
    return {
      name,
      content: "word",
      ...options,
    };
  } else if (
    type === "dict" ||
    type === "dict-furigana" ||
    type === "dict-kana"
  ) {
    const options: AnkiTemplateFieldWordOptionsV1 = {
      form: "dict-form",
      style: type === "dict-furigana" ?
        "furigana-anki" :
        type === "dict-kana" ?
        "kana-only" :
        "basic",
    };
    return { name, content: "word", ...options };
  } else if (
    type === "main-dict" ||
    type === "main-dict-furigana" ||
    type === "main-dict-kana"
  ) {
    const options: AnkiTemplateFieldWordOptionsV1 = {
      form: "main-dict-form",
      style: type === "main-dict-furigana" ?
        "furigana-anki" :
        type === "main-dict-kana" ?
        "kana-only" :
        "basic",
    };
    return {
      name,
      content: "word",
      ...options,
    };
  } else if (
    type === "sentence" ||
    type === "sentence-furigana" ||
    type === "sentence-kana" ||
    type === "sentence-cloze" ||
    type === "sentence-cloze-furigana"
  ) {
    const isCloze = type === "sentence-cloze" || type === "sentence-cloze-furigana";
    const options: AnkiTemplateFieldSentenceOptionsV1 = {
      style: type === "sentence-furigana" || type === "sentence-cloze-furigana" ?
        "furigana-anki" :
        type === "sentence-kana" ?
        "kana-only" :
        "basic",
      word: isCloze ? "cloze" : "bold",
    };
    return {
      name,
      content: "sentence",
      ...options,
    };
  } else if (
    type === "meaning" ||
    type === "meaning-full" ||
    type === "meaning-short"
  ) {
    return {
      ...newAnkiTemplateField(name, "meaning"),
      full_format: type === "meaning-short" ? "line" : "numbered",
    };
  } else if (
    type === "translated-sentence" ||
    type === "url" ||
    type === "link"
  ) {
    return {
      name,
      content: type,
    };
  } else {
    console.error(`Invalid Anki field template type '${type}' encountered for field: '${name}'`);
    return {
      name,
      content: "",
    };
  }
}

export function newAnkiTemplateField<C extends AnkiTemplateFieldContentV1>(
  name: string,
  content: C,
): AnkiTemplateFieldTypesV1[C];
export function newAnkiTemplateField(
  name: string,
  content: AnkiTemplateFieldContentV1,
): AnkiTemplateFieldV1 {
  if (
    content === "" ||
    content === "translated-sentence" ||
    content === "url" ||
    content === "link"
  ) {
    return {
      name,
      content: content,
    };
  } else if (content === "word") {
    return {
      name,
      content: content,
      form: "as-is",
      style: "basic",
    };
  } else if (content === "sentence") {
    return {
      name,
      content: content,
      style: "basic",
      word: "none",
    };
  } else if (content === "meaning") {
    return {
      name,
      content: content,
      full_format: "numbered",
      full_pos: false,
      full_max_item: 0,
      single_pos: false,
      single_max_item: 0,
    };
  } else {
    console.error(`Invalid Anki template field type '${content}'`);
    return {
      name,
      content: "",
    };
  }
}
