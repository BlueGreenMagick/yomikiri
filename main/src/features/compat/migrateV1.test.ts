import { expect, test } from "vitest";
import { migrateV1 } from "./migrateV1";
import type { StoredConfigurationV1 } from "./types/typesV1";

type StoredConfigurationVersion2 = StoredConfigurationV1 & { config_version: 2 };

const config2: StoredConfigurationVersion2 = {
  version: "0.1.3",
  "anki.enabled": true,
  "anki.connect_port": 8766,
  "anki.template": {
    deck: "Default",
    notetype: "Many Fields",
    fields: [
      {
        name: "Front",
        value: "word",
      },
      {
        name: "Back",
        value: "word-furigana",
      },
      {
        name: "3",
        value: "word-kana",
      },
      {
        name: "4",
        value: "dict",
      },
      {
        name: "5",
        value: "dict-furigana",
      },
      {
        name: "6",
        value: "dict-kana",
      },
      {
        name: "7",
        value: "main-dict",
      },
      {
        name: "8",
        value: "main-dict-furigana",
      },
      {
        name: "9",
        value: "main-dict-kana",
      },
      {
        name: "10",
        value: "sentence",
      },
      {
        name: "11",
        value: "sentence-furigana",
      },
      {
        name: "12",
        value: "sentence-kana",
      },
      {
        name: "13",
        value: "sentence-cloze",
      },
      {
        name: "14",
        value: "sentence-cloze-furigana",
      },
      {
        name: "15",
        value: "translated-sentence",
      },
      {
        name: "16",
        value: "meaning",
      },
      {
        name: "a",
        value: "meaning-full",
      },
      {
        name: "b",
        value: "meaning-short",
      },
      {
        name: "c",
        value: "url",
      },
      {
        name: "d",
        value: "link",
      },
    ],
    tags: "tag1 tag2",
  },
  "general.font_size": 16,
  config_version: 2,
};

test("migrate StoredConfigurationV1(config_version=2) to StoredConfigurationV2", () => {
  const { config } = migrateV1({ config: config2 });
  expect(config).toMatchSnapshot();
});
