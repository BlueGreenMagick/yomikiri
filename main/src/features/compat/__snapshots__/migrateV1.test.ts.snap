// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate StoredConfigurationV1(config_version=2) to StoredConfigurationV2 1`] = `
{
  "anki.anki_template": {
    "deck": "Default",
    "fields": [
      {
        "content": "word",
        "form": "as-is",
        "name": "Front",
        "style": "basic",
      },
      {
        "content": "word",
        "form": "as-is",
        "name": "Back",
        "style": "furigana-anki",
      },
      {
        "content": "word",
        "form": "as-is",
        "name": "3",
        "style": "kana-only",
      },
      {
        "content": "word",
        "form": "dict-form",
        "name": "4",
        "style": "basic",
      },
      {
        "content": "word",
        "form": "dict-form",
        "name": "5",
        "style": "furigana-anki",
      },
      {
        "content": "word",
        "form": "dict-form",
        "name": "6",
        "style": "kana-only",
      },
      {
        "content": "word",
        "form": "main-dict-form",
        "name": "7",
        "style": "basic",
      },
      {
        "content": "word",
        "form": "main-dict-form",
        "name": "8",
        "style": "furigana-anki",
      },
      {
        "content": "word",
        "form": "main-dict-form",
        "name": "9",
        "style": "kana-only",
      },
      {
        "content": "sentence",
        "name": "10",
        "style": "basic",
        "word": "bold",
      },
      {
        "content": "sentence",
        "name": "11",
        "style": "furigana-anki",
        "word": "bold",
      },
      {
        "content": "sentence",
        "name": "12",
        "style": "kana-only",
        "word": "bold",
      },
      {
        "content": "sentence",
        "name": "13",
        "style": "basic",
        "word": "cloze",
      },
      {
        "content": "sentence",
        "name": "14",
        "style": "furigana-anki",
        "word": "cloze",
      },
      {
        "content": "translated-sentence",
        "name": "15",
      },
      {
        "content": "meaning",
        "full_format": "numbered",
        "full_max_item": 0,
        "full_pos": false,
        "name": "16",
        "single_max_item": 0,
        "single_pos": false,
      },
      {
        "content": "meaning",
        "full_format": "numbered",
        "full_max_item": 0,
        "full_pos": false,
        "name": "a",
        "single_max_item": 0,
        "single_pos": false,
      },
      {
        "content": "meaning",
        "full_format": "line",
        "full_max_item": 0,
        "full_pos": false,
        "name": "b",
        "single_max_item": 0,
        "single_pos": false,
      },
      {
        "content": "url",
        "name": "c",
      },
      {
        "content": "link",
        "name": "d",
      },
    ],
    "notetype": "Many Fields",
    "tags": "tag1 tag2",
  },
  "anki.connect_port": 8766,
  "anki.enabled": true,
  "anki.template": {
    "deck": "Default",
    "fields": [
      {
        "name": "Front",
        "value": "word",
      },
      {
        "name": "Back",
        "value": "word-furigana",
      },
      {
        "name": "3",
        "value": "word-kana",
      },
      {
        "name": "4",
        "value": "dict",
      },
      {
        "name": "5",
        "value": "dict-furigana",
      },
      {
        "name": "6",
        "value": "dict-kana",
      },
      {
        "name": "7",
        "value": "main-dict",
      },
      {
        "name": "8",
        "value": "main-dict-furigana",
      },
      {
        "name": "9",
        "value": "main-dict-kana",
      },
      {
        "name": "10",
        "value": "sentence",
      },
      {
        "name": "11",
        "value": "sentence-furigana",
      },
      {
        "name": "12",
        "value": "sentence-kana",
      },
      {
        "name": "13",
        "value": "sentence-cloze",
      },
      {
        "name": "14",
        "value": "sentence-cloze-furigana",
      },
      {
        "name": "15",
        "value": "translated-sentence",
      },
      {
        "name": "16",
        "value": "meaning",
      },
      {
        "name": "a",
        "value": "meaning-full",
      },
      {
        "name": "b",
        "value": "meaning-short",
      },
      {
        "name": "c",
        "value": "url",
      },
      {
        "name": "d",
        "value": "link",
      },
    ],
    "notetype": "Many Fields",
    "tags": "tag1 tag2",
  },
  "config_version": 3,
  "general.font_size": 16,
  "version": "0.1.3",
}
`;
